### ISSUE TITLE
[Feature Request] Allow extended Vue constructors as mixins option

### ISSUE THREAD

**Issue by ktsn** [vuejs/vue#2956]
I'm using vue-class-component on my project and try to split component features to some small mixins.

I think it is good if the mixins option also allows extended Vue constructors because we can write components and mixins in uniformed ES2015 format.

In addition, there are an advantage of type system if we use with TypeScript.
For example, there are two mixins that receive props of event handlers to detect changing form values.

```js
@Component({
  props: {
    onChangeName: Function,
    onChangeEmail: Function
  }
})
class UserFormMixin extends Vue {
  onChangeName: (name: string) => void;
  onChangeEmail: (email: string) => void;
}

@Component({
  props: {
    onChangeHobby: Function,
    onChangeWebSite: Function
  }
})
class ProfileFormMixin extends Vue {
  onChangeHobby: (hobby: string) => void;
  onChangeWebSite: (webSite: string) => void;
}
```

These mixins can be applied to a component in following way if mixins option allows extended Vue constructors.
The `UserProfileForm` component must have the methods on `UserFormMixin` and `ProfileFormMixin`, and the component will receive the methods from its props.

```js
@Component({
  mixins: [
    UserFormMixin,
    ProfileFormMixin
  ]
})
class UserProfileForm extends Vue implements UserFormMixin, ProfileFormMixin {
  onChangeName: (name: string) => void;
  onChangeEmail: (email: string) => void;
  onChangeHobby: (hobby: string) => void;
  onChangeWebSite: (webSite: string) => void;
}
```

I would like to contribute this if this suggestion is ok :)

**Comment by yyx990803** [comment]
This should be fairly simple under the hood by detecting a function mixin
and use its options property instead.

**Comment by njleonzhang** [comment]
@ktsn I think this feature is great. especially the benefit to use type system. yyx990803 claim that this can be done easily, can you provide a sample?

**Comment by ktsn** [comment]
It's already implemented #2957

**Comment by njleonzhang** [comment]
@ktsn I need the integrate with type system. Can you check this question? 
https://forum.vuejs.org/t/typescript-to-restrict-vue-component-instance-must-have-special-properties-or-methods/21709/2
